File: detection/semgrep/fixtures/csharp/unscoped-query-object-fetch.cs
// Test fixture for rule: mcp-unscoped-query-object-fetch-csharp
//
// The query-scoped shape: an EF Core fetch where the tenant key is either
// present in the query (safe) or omitted (the S7 bug). The rule fires on the
// omission and stays silent when the tenant key is bound.
//
// Corpus signatures (5 total, all POOL):
// - daohainam/microservice-patterns: BookService Books.FindAsync(id)
// - daohainam/microservice-patterns: PaymentService Cards.FindAsync(id)
// - Azure-Samples/eShopLite 06-mcp: MemoryContext db.FindAsync<Product>(id) ← MCP server
// - TrashMob: SyncQueue db.FindAsync<PendingRouteSession>(sessionId)
// - (variant) FirstOrDefaultAsync also present as by-id fetch in same repos
//
// Two-way canary axis: is the fetch narrowed by an owner/tenant? (present ⇒ SILENT)

using System;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;

// FIRE: Find by id only (no tenant key) — EF Core DbSet.FindAsync / DbContext.FindAsync
public async Task<Credential> GetCredentialVuln(string id) {
    // ruleid: mcp-unscoped-query-object-fetch-csharp
    return await _context.Credentials.FindAsync(id);
}

// FIRE: DbContext.FindAsync<T>(id) — eShopLite 06-mcp pattern
public async Task<Product> GetProductVuln(string id) {
    // ruleid: mcp-unscoped-query-object-fetch-csharp
    return await _context.FindAsync<Product>(id);
}

// FIRE: FirstOrDefault with only id
public async Task<Note> GetNoteVuln(string id) {
    // ruleid: mcp-unscoped-query-object-fetch-csharp
    return await _context.Notes.FirstOrDefaultAsync(n => n.Id == id);
}

// FIRE: SingleOrDefault with only id
public async Task<Note> GetNoteSingleVuln(string id) {
    // ruleid: mcp-unscoped-query-object-fetch-csharp
    return await _context.Notes.SingleOrDefaultAsync(n => n.Id == id);
}

// FIRE: Remove with Find only
public async Task DeleteRecordVuln(string id) {
    // ruleid: mcp-unscoped-query-object-fetch-csharp
    var entity = await _context.Notes.FindAsync(id);
    if (entity != null) _context.Notes.Remove(entity);
}

// OK: Find by id with tenant key in predicate (SILENT - owner/tenant scoped)
public async Task<Credential> GetCredentialSafe(string id, string workspaceId) {
    // ok: mcp-unscoped-query-object-fetch-csharp
    return await _context.Credentials
        .FirstOrDefaultAsync(c => c.Id == id && c.WorkspaceId == workspaceId);
}

// OK: FirstOrDefault with orgId
public async Task<Note> GetNoteSafe(string id, string orgId) {
    // ok: mcp-unscoped-query-object-fetch-csharp
    return await _context.Notes
        .FirstOrDefaultAsync(n => n.Id == id && n.OrgId == orgId);
}

// OK: Remove with tenant key
public async Task DeleteRecordSafe(string id, string projectId) {
    // ok: mcp-unscoped-query-object-fetch-csharp
    var entity = await _context.Notes
        .FirstOrDefaultAsync(n => n.Id == id && n.ProjectId == projectId);
    if (entity != null) _context.Notes.Remove(entity);
}

// SILENT: Where clause carries tenant — .Where(x => x.OrgId == orgId).FirstOrDefaultAsync(x => x.Id == id)
public async Task<Note> GetNoteWhereSafe(string id, string orgId) {
    // ok: mcp-unscoped-query-object-fetch-csharp
    return await _context.Notes
        .Where(x => x.OrgId == orgId)
        .FirstOrDefaultAsync(x => x.Id == id);
}

class Credential { public string Id { get; set; } = ""; public string WorkspaceId { get; set; } = ""; }
class Product { public string Id { get; set; } = ""; public string TenantId { get; set; } = ""; }
class Note { public string Id { get; set; } = ""; public string OrgId { get; set; } = ""; public string ProjectId { get; set; } = ""; public string TenantId { get; set; } = ""; }
class AppDbContext : DbContext {
    public DbSet<Credential> Credentials { get; set; }
    public DbSet<Product> Products { get; set; }
    public DbSet<Note> Notes { get; set; }
}
AppDbContext _context = null;

File: detection/semgrep/fixtures/php/unscoped-query-object-fetch.php
<?php
// Test fixture for rule: mcp-unscoped-query-object-fetch-php
//
// The query-scoped shape: a model/repo fetch where the tenant key is either
// present in the query (safe) or omitted (the S7 bug). The rule fires on the
// omission and stays silent when the tenant key is bound.
//
// Corpus signatures (4 total):
// - POOL Contao/ProcessWire: ArticleModel::findBy(['id=?'], [$toolInputId])
// - POOL Contao: find(string $type, string $id, ...) MCP tool takes raw $id
// - POOL ProcessWire: findPages(array $input) MCP tool
// - ECO Eloquent: User::find($request->id), Team::findOrFail($id) — confirmed in ecosystem only
//
// NOTE: PHP MCP pool uses Contao Models (Model::findBy) and ProcessWire selectors,
// NOT Laravel Eloquent. Eloquent find/findOrFail is confirmed only in ecosystem.
// Pattern BOTH families.
//
// Two-way canary axis: is a tenant/owner column in the criteria? (present ⇒ SILENT)

use Illuminate\Database\Eloquent\Model;

// FIRE: Eloquent find by id only (ecosystem)
function getCredentialVuln(string $id): ?Model {
    // ruleid: mcp-unscoped-query-object-fetch-php
    return Credential::find($id);
}

// FIRE: Eloquent findOrFail by id only (ecosystem)
function getCredentialVulnFail(string $id): ?Model {
    // ruleid: mcp-unscoped-query-object-fetch-php
    return Credential::findOrFail($id);
}

// FIRE: Contao findByPk (POOL)
function getContaoVuln(string $id): ?Model {
    // ruleid: mcp-unscoped-query-object-fetch-php
    return ArticleModel::findByPk($id);
}

// FIRE: Contao findBy with id-only criteria (POOL - MCP tool)
function getContaoFindByVuln(array $input): ?Model {
    // ruleid: mcp-unscoped-query-object-fetch-php
    $id = $input['id'] ?? '';
    return ArticleModel::findBy(['id=?'], [$id]);
}

// FIRE: ProcessWire findPages (POOL - MCP tool takes raw id/selector)
function getProcessWireVuln(array $input): array {
    // ruleid: mcp-unscoped-query-object-fetch-php
    return PagesTools::findPages($input);
}

// FIRE: Doctrine find by id (if Doctrine repo)
function getDoctrineVuln(string $id): ?Model {
    // ruleid: mcp-unscoped-query-object-fetch-php
    return $repo->find($id);
}

// OK: Eloquent where with tenant key (SILENT - ecosystem)
function getCredentialSafe(string $id, string $workspace_id): ?Model {
    // ok: mcp-unscoped-query-object-fetch-php
    return Credential::where('id', $id)->where('workspace_id', $workspace_id)->first();
}

// OK: Contao findOneBy with tenant column (SILENT)
function getContaoSafe(string $id, string $tenant_id): ?Model {
    // ok: mcp-unscoped-query-object-fetch-php
    return ArticleModel::findOneBy(['id' => $id, 'tenant_id' => $tenant_id]);
}

// OK: Eloquent whereBelongsTo (SILENT - spatie/laravel-query-builder pattern)
function getNoteBelongsTo(string $id, $owner): ?Model {
    // ok: mcp-unscoped-query-object-fetch-php
    return Note::whereBelongsTo($owner)->where('id', $id)->first();
}

// OK: Doctrine with tenant criteria
function getDoctrineSafe(string $id, string $tenant_id): ?Model {
    // ok: mcp-unscoped-query-object-fetch-php
    return $repo->findOneBy(['id' => $id, 'tenant_id' => $tenant_id]);
}

// SILENT: Eloquent destroy with tenant check
function deleteRecordSafe(string $id, string $project_id): ?Model {
    // ok: mcp-unscoped-query-object-fetch-php
    $note = Note::where('id', $id)->where('project_id', $project_id)->first();
    return $note ? $note->delete() : null;
}

class Credential extends Model { }
class Note extends Model { }
class ArticleModel extends Model { 
    public static function findBy($cols, $vals) { return null; }
    public static function findByPk($id) { return null; }
    public static function findOneBy(array $criteria) { return null; }
}
class PagesTools { 
    public static function findPages(array $input) { return []; }
}
$repo = null;